internal/graph/graph.go

```diff
@@ -52,120 +50,101 @@ type ResolutionMetadata struct {
+	// a read of tuples and this one will not.
+	ComputedUsersetEdge
 )
 
-func (r RelationshipIngressType) String() string {
+func (r RelationshipEdgeType) String() string {
 	switch r {
-	case DirectIngress:
+	case DirectEdge:
 		return "direct"
-	case ComputedUsersetIngress:
+	case ComputedUsersetEdge:
 		return "computed_userset"
-	case TupleToUsersetIngress:
+	case TupleToUsersetEdge:
 		return "ttu"
 	default:
 		return "undefined"
 	}
 }
 
-type IngressCondition int
+type EdgeCondition int
 
 const (
 
-	// RequiresFurtherEvalCondition indicates an ingress condition whereby results expanded
-	// under such an ingress require further Check evaluation before a determination of the
-	// outcome can be made.
+	// RequiresFurtherEvalCondition indicates an edge condition whereby results found with ReverseExpandQuery
+	// require further Check evaluation before a determination of the outcome can be made.
 	//
 	// Relationships involving intersection ('and') and/or exclusion ('but not') fall under
-	// ingresses with this condition.
-	RequiresFurtherEvalCondition IngressCondition = iota
+	// edges with this condition.
+	RequiresFurtherEvalCondition EdgeCondition = iota
 
-	// NoFurtherEvalCondition indicates an ingress condition whereby results are factual and
+	// NoFurtherEvalCondition indicates an edge condition whereby results found with ReverseExpandQuery are factual and
 	// known to be true and require no further evaluation before a determination of the outcome
 	// can be made.
 	NoFurtherEvalCondition
 )
 
-// RelationshipIngress represents a possible ingress point between some source object reference
-// and a target user reference.
-type RelationshipIngress struct {
+// RelationshipEdge represents a possible relationship some source object reference
+// and a target user reference. The possibility is realized depending on the tuples and on the edge's type.
+type RelationshipEdge struct {
+	Type RelationshipEdgeType
 
-	// The type of the relationship ingress
-	Type RelationshipIngressType
+	// The edge is directed towards this node, which can be like group:*, or group, or group:member
+	TailUsersetNode *openfgav1.RelationReference
 
-	// The relationship reference that defines the ingress to some target relation
-	Ingress *openfgav1.RelationReference
-
-	// TuplesetRelation defines the tupleset relation reference that relates a source
-	// object reference with a target if the type of the relationship ingress is that
-	// of a TupleToUserset
-	// TODO this can be just a string for the relation (since the type will be the same as Ingress.Type)
+	// If the type is TupleToUsersetEdge, this defines the TTU condition
+	// TODO this can be just a string for the relation (since the type will be the same as TailUsersetNode.Type)
 	TuplesetRelation *openfgav1.RelationReference
 
-	Condition IngressCondition
+	// TODO this is leaking implementation details of ReverseExpand. This can be a boolean saying
```

Comment:
Not sure I understand the motivation behind this comment?

Are you suggesting that instead of an `EdgeCondition` being `RequiresFurtherEvalCondition` or `NoFurtherEvalRequired` (which is leaky), that this should just change to boolean type here?

Reply:
Yes

> Not sure I understand the motivation behind this comment?

It would separate concerns. Right now, `graph` knows a bit of how ReverseExpand works. I want `graph` to just tell me whether a node is `and` or `but not`.

And then in ReverseExpand we can do

```go
if (any of the edges we visited so far have this boolean = true) {
  resultStatus = RequiresFurtherEvalStatus
}

Reply:
Gotcha. Makes sense.